### PR TITLE
fix: users logged out after SDK upgrade due to different cache path

### PR DESCRIPTION
**Pull request by rommansabbir**
ParsePlugins ::
- `getCacheDir()`, `getFilesDir()` APIs now has support for backward compatibility before migration to v3.0.0

### CODE REVIEW

File: parse/src/main/java/com/parse/ParsePlugins.java
Comment (L3K0V) [inline parse/src/main/java/com/parse/ParsePlugins.java:188]: I remember that because of the Android's JVM specifics and the application lifecycle it's not good to use `deleteOnExit()`. This will properly work on desktop JVM but on Android it's questionable. 

- https://groups.google.com/g/android-developers/c/2FOEz35k_Uc
- https://developer.android.com/reference/java/io/File#deleteOnExit()

> Note that on Android, the application lifecycle does not include VM termination, so calling this method will not ensure that files are deleted. Instead, you should use the most appropriate out of:

>  - Use a finally clause to manually invoke [delete()](https://developer.android.com/reference/java/io/File#delete()).
> - Maintain your own set of files to delete, and process it at an appropriate point in your application's lifecycle.
> - Use the Unix trick of deleting the file as soon as all readers and writers have opened it. No new readers/writers will be able to access the file, but all existing ones will still have access until the last one closes the file.

File: parse/src/main/java/com/parse/ParsePlugins.java
Comment (L3K0V) [inline parse/src/main/java/com/parse/ParsePlugins.java:183]: I can see that you copy reference, but what's the idea behind @rommansabbir? Coping reference will no do the migration itself, right?

Where the actual data copy is done from the old to the new path aka. the migration?

**Review by L3K0V — CHANGES_REQUESTED** [review]

**Comment by L3K0V** [comment]
Here you can see an migration logic about the push notifications, I have expected to see something similar but much more simple: https://github.com/parse-community/Parse-SDK-Android/commit/7d0faa38bf3ccc85b715dc61965d5b82103c19a4#diff-6cb14853da236c30db3f14e4005ea40c5c45cb8fcb9fc49541693186620de1dcR60

File: parse/src/main/java/com/parse/ParsePlugins.java
Comment (mtrezza) [inline parse/src/main/java/com/parse/ParsePlugins.java:200]: Could you please deduplicate this logic into a separate method?

**Review by mtrezza — CHANGES_REQUESTED** [review]

**Comment by rommansabbir** [comment]
@mtrezza @mman @L3K0V First of all, sorry for messed up with the PR. This is my first contribution to a such big community of Parse.

About my logic for cache path migrations. Based on @mman migration thoughts (https://github.com/parse-community/Parse-SDK-Android/issues/1158#issuecomment-1110097992)

I came up with this ::

- First, looking for the all files (including sub directory) in the old "app_Parse" directory
- If files exists, moving the all files to the directory "com.parse" by following the old sub directory name. After that delete the old file (already copied) silently.
- For each file, migration will be failed if there is already a file exist in the new directory with the same file name or other known cases.

If the solution seems suitable for migration process, I will create a new PR (which will include changes in `ParsePlugins.java` and `ParseFileUtils.java`).

Let me know, Thanks.

**Comment by mman** [comment]
@rommansabbir Yes, this is the approach I would take, but please note that since 3.0.0 is already in the wild, and since it forced people to re-login, there will be cases where there will be fresh files in new location `com.parse` as well as files in old location, the `app_Parse` directory. In that case, the new files should be used, and old just deleted...

**Comment by mman** [comment]
@rommansabbir That is exactly what I have not yet had a time to investigate properly and in detail (read: This PR is actually trickier then it may appear on a first look).

**Comment by rommansabbir** [comment]
@mman I think we can find out from where the old API is accessed and which `FILE_NAME` (File name) was used, eg. `File(ParsePlugins.get().getParseDir(), FILENAME_CURRENT_CONFIG)`. We can look for others `Constants` like `FILENAME_CURRENT_CONFIG`. So, when migration is ongoing, we can check for item name matches with our `Constants` list or not, if match - then move item to `*/files/com.parse/` else `*/cache/com.parse/`

**Comment by mtrezza** [comment]
> So, when running migration, for a specific file, in which location I should point to move the spcific file, */cache/com.parse or */files/com.parse?.

The two storages have a different purpose and are managed differently. That determines which file should go where.
- The cache dir is authoritatively managed by the OS and only co-managed by the app. It must not be considered persistent, in other words, the data the app stores there can be deleted by the OS at any time. That also means that if the app doesn't properly delete files the OS will make sure the cache doesn't swell up too much.
- The files dir is exclusively managed by the app. It is persistent and *must* be managed by the app. If the app doesn't delete files there, the files will stay and the app's footprint will swell up until the OS will report a storage issue to the user.

**Comment by rommansabbir** [comment]
````
ParsePlugins::installationId(){File(getParseDir(), INSTALLATION_ID_LOCATION)}
ParseCorePlugins::getCurrentUserController() {File(Parse.getParseDir(), FILENAME_CURRENT_USER)}
ParseCorePlugins::getConfigController() {File(ParsePlugins.get().getParseDir(), FILENAME_CURRENT_CONFIG)}
ParseCorePlugins::getCurrentInstallationController() {File(ParsePlugins.get().getParseDir(), FILENAME_CURRENT_INSTALLATION)}
ParseCommandCache::getCacheDir(){File(Parse.getParseDir(), "CommandCache")}
LocalIdManager(Parse.getParseDir()){File(root, "LocalId")}
PushRouter::getInstance(){File(ParsePlugins.get().getParseDir(), LEGACY_STATE_LOCATION)}
````
from where `ParsePlugins.get().getParseDir()` API has been accessed. All of those files managed by SDK itself, so when migration is ongoing, we can look for those listed files and move files to `*/files/com.parse/` directory, then delete the old files from `app_Parse` by covering known cases.

After that, all other files that left in `app_Parse` directory should move to `*/cache/com.parse/` directory by covering known cases.


> * The files dir is exclusively managed by the app. It is persistent and _must_ be managed by the app. If the app doesn't delete outdated files there the files will stay and the app's footprint will swell up until the OS will report a storage issue to the user.

Version: 2.0.6
````
ParseCorePlugins::getFileController(){
             ParseFileController(
                    ParsePlugins.get().restClient(), 
                    Parse.getParseCacheDir("files")
            ))
}
````
calling `Parse.getParseCacheDir("files")` API which point to `*/cache/com.parse/`. In this case, all files (eg. when storing an `ParseObject` to `LocalDataStore` that contain an _File_ itself call `something.mp3`) that are stored in the cache directory should point to `*/files/com.parse/` as  `something.mp3` is managed by SDK itself.

**Comment by mtrezza** [comment]
> All of those files managed by SDK itself, so when migration is ongoing, we can look for those listed files and move files to */files/com.parse/ directory, then delete the old files from app_Parse by covering known cases.

Sounds good

> After that, all other files that left in app_Parse directory should move to */cache/com.parse/ directory by covering known cases.

What are these other files?

How about the other modules like facebook, fcm, it is possible that they store anything in these dirs?

**Comment by rommansabbir** [comment]
@mtrezza I found those usages in the mentioned list. If any other module deal with _File_, at the end they all point to the same API `ParsePlugins.get().getParseDir()` to retreive `File` object.

> What are these other files?

I'm not sure about how many files (`File`) is accessed (CRUD) to old dir (`app_Parse`) by the SDK itself. So,
- Files that matches with our lists, will be moved to `*/files/com.parse/` but what about others files in the old dir? (If exists)
- We can assume those files should be moved to `*/cache/com.parse/` dir.

**Comment by mtrezza** [comment]
I wonder what these other files could be? I assume it's only the Parse SDK itself that writes to that dir, so any write logic, even for temporary files should be identifiable in code. Have you seen that there are other files, or is that just an assumption?

**Comment by mtrezza** [comment]
OK, so I guess let's go with your suggested migration concept and then just test it out.

**Comment by rommansabbir** [comment]
> You mean pre 3.x or even with the latest version of the SDK?

pre 3.x since, from 3.x API removed from codebase and using `*/files/com.parse/` and `*/cache/com.parse/`.

Also, Can I write the test in Kotlin?

**Comment by mtrezza** [comment]
Not sure about Kotlin, we probably want to keep all tests in a consistent language, what do you think @L3K0V?

**Comment by L3K0V** [comment]
Tests should be in Java for now.

**Comment by mtrezza** [comment]
@rommansabbir you get a 403 error pushing to the branch in your own fork?

**Comment by rommansabbir** [comment]
> you get a 403 error pushing to the branch in your own fork? If so, you need to authenticate yourself for GitHub to accept a push.

I think, Git was trying push to the remote branch as well as my forked branch. But yeah, code is psuhed to my fork's branch.

I have added test for the migration process. Still, one more thing left, calling new overloaded APIs `ParsePlugins.get().getCacheDir(true)` and `ParsePlugins.get().getFilesDir(true)` should be used instead of old APIs in the respective classes.

**Comment by rommansabbir** [comment]
Thanks, I will definitely ask you for further guidance. Can review the changes? So that, we can close this ASAP.

**Comment by mtrezza** [comment]
> it will be called everytimes, when cache or files API is called

I would call the migration routine only once on SDK initialization.

File: parse/src/main/java/com/parse/ParsePlugins.java
Comment (L3K0V) [inline parse/src/main/java/com/parse/ParsePlugins.java:244]: Is there any reason this to be public?

File: parse/src/main/java/com/parse/ParsePlugins.java
Comment (L3K0V) [inline parse/src/main/java/com/parse/ParsePlugins.java:254]: Why is public and why is called replace last, but we have text replace first?

File: parse/src/main/java/com/parse/ParsePlugins.java
Comment (L3K0V) [inline parse/src/main/java/com/parse/ParsePlugins.java:281]: For readability this can be converted to something like:

```java
String[] strArray = {"installationId", "currentUser", "currentConfig", "currentInstallation", ...};
List<String> configFilesToMigrateToFilesDir = Arrays.asList(strArray);
```

File: parse/src/main/java/com/parse/ParsePlugins.java
Comment (L3K0V) [inline parse/src/main/java/com/parse/ParsePlugins.java:302]: Can we add log somewhere here what was migrated? I believe it will be interesting in a verbose mode for the developers.

File: parse/src/main/java/com/parse/ParsePlugins.java
Comment (rommansabbir) [reply parse/src/main/java/com/parse/ParsePlugins.java:244]: No, I don't see any valid to be public. Also, Migration process is now independent module and protected, so no more public access.

File: parse/src/main/java/com/parse/ParsePlugins.java
Comment (rommansabbir) [reply parse/src/main/java/com/parse/ParsePlugins.java:254]: Code refactored and reason is `Before returning the path, replace file name from the last, eg. dir name & file name could be same, as we want to get only dir name.`

File: parse/src/main/java/com/parse/ParsePlugins.java
Comment (rommansabbir) [reply parse/src/main/java/com/parse/ParsePlugins.java:281]: Thanks. Refactored.

File: parse/src/main/java/com/parse/ParsePlugins.java
Comment (rommansabbir) [reply parse/src/main/java/com/parse/ParsePlugins.java:302]: Yes, Added (`PLog.v()`) verbos log for events like _Copying_, _Already Exist_, _Deletion_, _Migration_Completion_.

**Review by rommansabbir — COMMENTED** [review]
fix: parse cache dir migration routine on SDK initialization (only once) (#1168)

**Review by L3K0V — APPROVED** [review]
LGTM, but it would be great if someone else take a look as well.

**Review by rommansabbir — COMMENTED** [review]
Fixed SpotlessJavaCheck Violations gradle build issue for PR (#1168)

**Comment by mtrezza** [comment]
> LGTM, but it would be great if someone else take a look as well.

Would be nice to have a pre-release branch to test things out easier, to consider for the future.

@PLR2388 Since you originally reported the issue, do you think you could test out this fix to see whether the migration happens successfully?

**Comment by rommansabbir** [comment]
@mman @mtrezza Any update on this PR?

**Comment by mtrezza** [comment]
I would like to have someone test this out with an actual app before we merge it, just to validate that it works.

**Comment by rommansabbir** [comment]
@mtrezza I will review the codes and will update you later.

**Comment by mman** [comment]
@rommansabbir I will try to test until the end of this week...

**Comment by rommansabbir** [comment]
@mtrezza According to the code comments:

- _Make sure the data on disk for Parse is for the current application_ on Parse SDK Init.
- _Verifies that the data stored on disk for Parse was generated using the same application that is running now_
- _Make sure the current version of the cache is for this application id else the applicationId file was malformed or something_
- _The application id has changed, so everything on disk is invalid._

**Comment by mtrezza** [comment]
Thanks for investigating, so the `applicationId` there is essentially to validate that the cache belongs to the app. That makes sense I guess. So if an app would allow to switch between different application IDs, the cache would get invalidated.

**Comment by mtrezza** [comment]
I can confirm that the migration works fine.

It moves from cache to files:
- currentConfig
- currentInstallation
- currentUser

It does not move from cache to files:
- installationId

Everything as expected, installation update and user requests also work fine.

**Comment by mtrezza** [comment]
Changelog entry:
> ### Fix
> - users logged out after SDK upgrade due to different cache path; this fixes the bug that was introduced with release 3.0.0 which ignores SDK-internal data that is stored locally on the client side (#1168)

**Comment by mtrezza** [comment]
@mman If you want to test that would be great, then we'll wait until the end of the week with the release.

**Review by mtrezza — APPROVED** [review]

**Comment by rommansabbir** [comment]
> Great job, @rommansabbir! Would you be interested in joining the Parse Android SDK team? No commitment in terms of contribution, but it would be great if you could at times help out with things like PR reviews for example.

Yes, I would love to 😃

**Comment by mman** [comment]
@rommansabbir @mtrezza I'd love to double check, the code looks good. What's the easiest way to build my app against this branch. I have tried naive approach to depend directly on git branch, but that apparently does not work. What steps are needed? Sorry for my lame question.

**Comment by rommansabbir** [comment]
image

Add this branch dependency:

`implementation 'com.github.rommansabbir:Parse-SDK-Android:patch-1-SNAPSHOT'`

Now, you can install the latest SDK from this branch.

**Comment by mtrezza** [comment]
You can also close the repo, checkout the PR branch, build to local maven and reference the builds.

**Comment by mman** [comment]
Thanks @rommansabbir, I was missing the `-SNAPSHOT` in the github reference.

```
implementation 'com.github.rommansabbir:Parse-SDK-Android:patch-1-SNAPSHOT'
```

I have now tried our apps and looks like this PR does correctly move all files around.

Thanks, LGTM 👍

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [3.0.1](https://github.com/parse-community/Parse-SDK-Android/releases/tag/3.0.1)

**Pull request merged**
